### ISSUE TITLE
False negative passing of recursive union types

### ISSUE THREAD

**Issue by jkb0o** [microsoft/pyright#1887]
**Describe the bug**
Pyright sometimes does not warn me about wrong types when using recursive union-types.

**To Reproduce**
```python
# test.py
from typing import Dict, List, Union

Json = Union[None, int, str, float, List['Json'], Dict[str, 'Json']]

class Obj(object):
    pass

a1: Json = {
    "a": 1,
    "b": Obj(), # Obj doesn't matches any Json type
}
```

`pyright test.py` passes without any warnings.

**Expected behavior**
I expected some error, for example, `Expression of type "dict[str, Any]" cannot be assigned to declared type "Json"`

**VS Code extension or command-line**
pyright cli tool version 1.1.141

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.142, which I just published. It will also be included in the next release of pylance.

**Comment by jkb0o** [comment]
Thanks!
